File: integration-tests/test_configuration.py
import subprocess

from testcase import TestCase
from testcase import CT_API_KEY
from testcase import CT_PROJ
from testcase import CT_ENV
from testcase import CT_PROFILE
from testcase import CT_REST_DEBUG
from testcase import CT_REST_PAGE_SIZE
from testcase import CT_REST_SUCCESS
from testcase import CT_TIMEOUT
from testcase import REDACTED
from testcase import find_by_prop


class TestConfiguration(TestCase):
    basic_prof_name = "cli-int-basic-prof-test"
    basic_child_prof = "cli-int-basic-child-test"
    current_prof_name = "cli-int-curr-prof-test"

    def tearDown(self) -> None:
        # clean up any stranded profiles
        for profile in [self.basic_prof_name, self.current_prof_name, self.basic_child_prof]:
            cmd = self.get_cli_base_cmd() + f"profile delete -y '{profile}'"
            subprocess.run(cmd, shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

        super().tearDown()

    def test_configuration_profile(self):
        base_cmd = self.get_cli_base_cmd()
        cmd_env = self.get_cmd_env()
        conf_cmd = base_cmd + "conf "
        list_cmd = conf_cmd + "prof ls "
        set_cmd = conf_cmd + "profile set "
        prof_name = self.basic_prof_name
        child_name = self.basic_child_prof
        desc1 = "Profile to use for something"
        desc2 = "alternate description"
        api_key1 = "bogus-key-value"
        env1 = "some-environment"
        proj1 = "my-proj-name"

        # make sure it is not already present
        result = self.run_cli(cmd_env, list_cmd)
        self.assertResultSuccess(result)
        self.assertNotIn(prof_name, result.out())

        # verify a good message for a missing profile
        result = self.run_cli(cmd_env, base_cmd + f"--profile '{prof_name}' env ls -v")
        self.assertResultError(result, f"Failed to load configuration from profile '{prof_name}'")

        # create it
        result = self.run_cli(cmd_env, set_cmd + f"'{prof_name}' -k '{api_key1}'")
        self.assertResultSuccess(result)
        self.assertIn(f"Created profile '{prof_name}'", result.out())

        # verify it was created
        result = self.run_cli(cmd_env, list_cmd)
        self.assertResultSuccess(result)
        self.assertIn(prof_name, result.out())

        # update some fields
        result = self.run_cli(cmd_env, set_cmd + f"'{prof_name}' -d '{desc1}' -e '{env1}'")
        self.assertResultSuccess(result)
        self.assertIn(f"Updated profile '{prof_name}'", result.out())

        # make sure API key is not shown, but other parameters are
        result = self.run_cli(cmd_env, list_cmd + "-vf csv")
        self.assertResultSuccess(result)
        self.assertIn(f"{prof_name},{REDACTED},{env1},,{desc1}", result.out())

        # update some fields
        result = self.run_cli(cmd_env, set_cmd + f"'{prof_name}' -p '{proj1}' -d '{desc2}'")
        self.assertResultSuccess(result)

        # make sure API key is not shown, but other parameters are
        result = self.run_cli(cmd_env, list_cmd + "-svf csv")
        self.assertResultSuccess(result)
        self.assertIn(f"{prof_name},{api_key1},{env1},{proj1},{desc2}", result.out())

        # update without any properties produces warning
        result = self.run_cli(cmd_env, set_cmd + f"'{prof_name}'")
        self.assertResultWarning(result, f"Nothing to change for profile '{prof_name}'")

        profiles = self.get_cli_entries(cmd_env, list_cmd + "-svf json", "profile")
        prof = [p for p in profiles if p.get("Name") == prof_name][0]
        self.assertEqual(api_key1, prof.get("API"))
        self.assertEqual(env1, prof.get("Environment"))
        self.assertEqual(proj1, prof.get("Project"))
        self.assertEqual(desc2, prof.get("Description"))

        # create a child
        child_project = "this-is-a-bogus-project-name"
        result = self.run_cli(cmd_env, set_cmd + f"'{child_name}' -s '{prof_name}' -p '{child_project}'")
        self.assertResultSuccess(result)

        # verify child was created
        result = self.run_cli(cmd_env, list_cmd + "-v -f csv")
        self.assertResultSuccess(result)
        self.assertIn(child_name, result.out())
        self.assertIn(child_project, result.out())

        # see that creating with a non-existent source/parent fails
        missing_child = "cli-int-prof-missing-child"
        new_child = "cli-int-prof-never-created"
        result = self.run_cli(cmd_env, set_cmd + f"'{new_child}' -s '{missing_child}'")
        self.assertResultError(result, f"Source profile '{missing_child}' does not exist")

        # delete the child
        result = self.run_cli(cmd_env, conf_cmd + f"p d -y '{child_name}'")
        self.assertResultSuccess(result)

        # delete it
        result = self.run_cli(cmd_env, conf_cmd + f"p d -y '{prof_name}'")
        self.assertResultSuccess(result)
        self.assertIn(f"Deleted profile '{prof_name}'", result.out())

        # verify it is gone
        result = self.run_cli(cmd_env, list_cmd)
        self.assertResultSuccess(result)
        self.assertNotIn(prof_name, result.out())

        # see deletion is idempotent
        result = self.run_cli(cmd_env, conf_cmd + f"profile delete -y '{prof_name}'")
        self.assertResultWarning(result, f"Profile '{prof_name}' does not exist")

    def test_configuration_current(self):
        base_cmd = self.get_cli_base_cmd()
        cmd_env = self.get_cmd_env()
        conf_cmd = base_cmd + "conf "
        curr_cmd = conf_cmd + "curr "

        prof_name = self.current_prof_name
        api_key = "bogus-key-value"
        env_name = "some-environment"
        proj_name = "my-proj-name"
        set_cmd = conf_cmd + f"prof set '{prof_name}' "

        result = self.run_cli(cmd_env, curr_cmd)
        self.assertResultSuccess(result)
        orig_env = result.out()

        # create the profile
        result = self.run_cli(cmd_env, set_cmd + f"-k '{api_key}' -e '{env_name}' -p '{proj_name}'")
        self.assertResultSuccess(result)

        # nothing changed, since did not setup to use new profile
        result = self.run_cli(cmd_env, curr_cmd)
        self.assertResultSuccess(result)
        self.assertEqual(orig_env, result.out())

        # now, set the environment to use the profile, and remove other environmental stuff
        cmd_env[CT_PROFILE] = prof_name
        cmd_env.pop(CT_API_KEY, None)
        cmd_env.pop(CT_PROJ, None)
        cmd_env.pop(CT_ENV, None)

        # see that things have changed
        result = self.run_cli(cmd_env, curr_cmd)
        self.assertResultSuccess(result)
        self.assertNotEqual(orig_env, result.out())
        self.assertIn(prof_name, result.out())

        # check the "all" parameters
        profile = self.get_cli_entries(cmd_env, curr_cmd + "-sf json", "profile")
        this_profile = f"profile ({prof_name})"
        param_names = [e.get("Parameter") for e in profile]
        expected = ["Profile", "API key", "Organization", "User", "Role", "Project", "Environment"]
        self.assertEqual(param_names, expected)

        entry = find_by_prop(profile, "Parameter", "Profile")[0]
        self.assertEqual(entry.get("Value"), prof_name)
        self.assertEqual(entry.get("Source"), "shell")

        entry = find_by_prop(profile, "Parameter", "API key")[0]
        self.assertEqual(entry.get("Value"), api_key)
        self.assertEqual(entry.get("Source"), this_profile)

        entry = find_by_prop(profile, "Parameter", "Project")[0]
        self.assertEqual(entry.get("Value"), proj_name)
        self.assertEqual(entry.get("Source"), this_profile)

        entry = find_by_prop(profile, "Parameter", "Environment")[0]
        self.assertEqual(entry.get("Value"), env_name)
        self.assertEqual(entry.get("Source"), this_profile)

        entry = find_by_prop(profile, "Parameter", "Organization")[0]
        self.assertEqual(entry.get("Value"), "")
        self.assertEqual(entry.get("Source"), "")

        entry = find_by_prop(profile, "Parameter", "User")[0]
        self.assertEqual(entry.get("Value"), "")
        self.assertEqual(entry.get("Source"), "")

        entry = find_by_prop(profile, "Parameter", "Role")[0]
        self.assertEqual(entry.get("Value"), "")
        self.assertEqual(entry.get("Source"), "")

        ##############################
        # extended version
        cmd_env[CT_REST_SUCCESS] = "a,b,c,d"
        cmd_env[CT_REST_DEBUG] = "false"
        cmd_env[CT_REST_PAGE_SIZE] = "9"
        cmd_env[CT_TIMEOUT] = "33"
        profile = self.get_cli_entries(cmd_env, curr_cmd + "-xsf json", "profile")
        this_profile = f"profile ({prof_name})"
        param_names = [e.get("Parameter") for e in profile]
        extended = [
            "Profile",
            "API key",
            "Organization",
            "User",
            "Role",
            "Project",
            "Environment",
            "CLI version",
            "Server URL",
            "Request timeout",
            "REST debug",
            "REST success",
            "REST page size",
            "Accept Invalid Certs",
        ]
        self.assertEqual(param_names, extended)

        entry = find_by_prop(profile, "Parameter", "REST debug")[0]
        self.assertEqual(entry.get("Value"), "false")
        self.assertEqual(entry.get("Source"), "shell")

        entry = find_by_prop(profile, "Parameter", "REST success")[0]
        self.assertEqual(entry.get("Value"), "a, b, c, d")
        self.assertEqual(entry.get("Source"), "shell")

        entry = find_by_prop(profile, "Parameter", "REST page size")[0]
        self.assertEqual(entry.get("Value"), "9")
        self.assertEqual(entry.get("Source"), "shell")

        entry = find_by_prop(profile, "Parameter", "Request timeout")[0]
        self.assertEqual(entry.get("Value"), "33")
        self.assertEqual(entry.get("Source"), "shell")

        cmd_env.pop(CT_REST_SUCCESS)
        cmd_env.pop(CT_REST_DEBUG)
        cmd_env.pop(CT_REST_PAGE_SIZE)
        cmd_env.pop(CT_TIMEOUT)

        ##############################
        # test with command line arguments
        bogus_api_key = "not-a-real-api-key"
        cmd = base_cmd + f"--api-key '{bogus_api_key}' --profile '{prof_name}' conf curr -sf json"
        profile = self.get_cli_entries(cmd_env, cmd, "profile")

        entry = find_by_prop(profile, "Parameter", "API key")[0]
        self.assertEqual(entry.get("Value"), bogus_api_key)
        self.assertEqual(entry.get("Source"), "argument")

        entry = find_by_prop(profile, "Parameter", "Profile")[0]
        self.assertEqual(entry.get("Value"), prof_name)
        self.assertEqual(entry.get("Source"), "argument")

        entry = find_by_prop(profile, "Parameter", "Organization")[0]
        self.assertEqual(entry.get("Value"), "")
        self.assertEqual(entry.get("Source"), "")

        entry = find_by_prop(profile, "Parameter", "User")[0]
        self.assertEqual(entry.get("Value"), "")
        self.assertEqual(entry.get("Source"), "")

        entry = find_by_prop(profile, "Parameter", "Role")[0]
        self.assertEqual(entry.get("Value"), "")
        self.assertEqual(entry.get("Source"), "")

        ##############################
        # delete the profile
        result = self.run_cli(cmd_env, conf_cmd + f"p d -y '{prof_name}'")
        self.assertResultSuccess(result)

        # when profile is not found, command succeeds without the bits from the config
        result = self.run_cli(cmd_env, curr_cmd + "-s")
        self.assertResultSuccess(result)
        self.assertIn(prof_name, result.out())
        self.assertNotIn(api_key, result.out())
        self.assertNotIn(env_name, result.out())
        self.assertNotIn(proj_name, result.out())
        self.assertNotIn(this_profile, result.out())

        ##############################
        # back to the original user values  -- do not show secrets
        cmd_env = self.get_cmd_env()
        profile = self.get_cli_entries(cmd_env, base_cmd + "conf curr -f json", "profile")
        entry = find_by_prop(profile, "Parameter", "API key")[0]
        self.assertEqual(entry.get("Value"), REDACTED)
        entry = find_by_prop(profile, "Parameter", "Organization")[0]
        self.assertNotEqual(entry.get("Value"), "")
        self.assertEqual(entry.get("Source"), "API key")
        entry = find_by_prop(profile, "Parameter", "User")[0]
        self.assertNotEqual(entry.get("Value"), "")
        self.assertEqual(entry.get("Source"), "API key")
        entry = find_by_prop(profile, "Parameter", "Role")[0]
        self.assertIn(entry.get("Value"), ["owner", "admin", "contrib"])
        self.assertEqual(entry.get("Source"), "API key")

File: tests/test_configuration.rs
use cloudtruth_config::{
    CT_PROFILE, CT_REQ_TIMEOUT, CT_REST_DEBUG, CT_REST_PAGE_SIZE, CT_REST_SUCCESS,
};
use cloudtruth_test_harness::output::profile::*;
use cloudtruth_test_harness::prelude::*;
use maplit::hashmap;

#[test]
#[use_harness]
fn test_configuration_profile() {
    // create test user
    let test_user = User::with_prefix("configuration-user").create();

    // test missing profile
    let missing_profile = Profile::with_prefix("missing-profile");
    cloudtruth!("--profile '{missing_profile}' env ls -v")
        .assert()
        .failure()
        .stderr(contains!(
            "Failed to load configuration from profile '{missing_profile}'"
        ));

    // create profile with test user API key
    let test_profile = Profile::with_prefix("configuration-profile")
        .api_key(test_user.api_key())
        .create();

    // verify profile exists
    cloudtruth!("--profile '{test_profile}' configuration profile list")
        .assert()
        .success()
        .stdout(contains(test_profile.name()));

    // update some profile fields
    cloudtruth!("conf prof set '{test_profile}' -d 'Test Profile' -e 'config-test-environment'")
        .assert()
        .success()
        .stdout(contains!("Updated profile '{test_profile}'"));

    // make sure API key is not shown, but other parameters are
    cloudtruth!("conf prof ls -vf csv")
        .assert()
        .success()
        .stdout(contains!(
            "{test_profile},*****,config-test-environment,,Test Profile"
        ));

    // update some more fields
    cloudtruth!(
        "conf prof set '{test_profile}' -p 'config-test-project' -d 'Updated Test Profile'"
    )
    .assert()
    .success();

    // show all parameters including secrets
    cloudtruth!("conf prof ls -vsf csv")
        .assert()
        .success()
        .stdout(contains!(
        "{test_profile},{api_key},config-test-environment,config-test-project,Updated Test Profile",
        api_key = test_user.api_key()
    ));

    // update without any options produces warning
    cloudtruth!("conf prof set '{test_profile}'")
        .assert()
        .success()
        .stderr(contains!("Nothing to change for profile '{test_profile}'"));

    // test JSON
    cloudtruth!("conf prof list -vsf json")
        .assert()
        .success()
        .stdout(json(prop(
            "profile",
            find_entry(
                prop("Name", value(test_profile.name().as_str())),
                all!(
                    prop("API", value(test_user.api_key())),
                    prop("Environment", value("config-test-environment")),
                    prop("Project", value("config-test-project")),
                    prop("Description", value("Updated Test Profile"))
                ),
            ),
        )));

    // create child profile
    let child_proj = Project::with_prefix("config-test-child-project");
    let child_profile = Profile::with_prefix("config-child-profile")
        .source(&test_profile)
        .project(&child_proj)
        .create();

    // verify child was created
    cloudtruth!("conf prof ls -v -f csv")
        .assert()
        .success()
        .stdout(contains_all([child_profile.name(), child_proj.name()]));

    // try to create child with non-existence parent
    let invalid_child_profile = Profile::with_prefix("config-missing-parent");
    let missing_profile = Profile::with_prefix("missing-profile");
    cloudtruth!("conf prof set '{invalid_child_profile}' -s '{missing_profile}'")
        .assert()
        .failure()
        .stderr(contains!(
            "Source profile '{missing_profile}' does not exist"
        ));

    // delete profiles
    drop(child_profile);
    let deleted_profile = Profile::from_name(test_profile.name().clone());
    drop(test_profile);

    // verify it is gone
    cloudtruth!("conf prof ls")
        .assert()
        .success()
        .stdout(not(contains(deleted_profile.name())));

    // delete is idempotent
    cloudtruth!("config prof delete -y '{deleted_profile}'")
        .assert()
        .success()
        .stderr(contains!("Profile '{deleted_profile}' does not exist"));
}

#[test]
#[use_harness]
fn test_configuration_current() {
    let cmd = cloudtruth!("config current").assert().success();
    let orig_profile = Profile::from_string(String::from_utf8_lossy(&cmd.get_output().stdout));
    let test_proj = Project::with_prefix("current-profile-proj");
    let test_env = Environment::with_prefix("current-profile-env");
    let test_profile = Profile::with_prefix("current-profile")
        .api_key("bogus-key-value")
        .project(&test_proj)
        .env(&test_env)
        .create();

    // current profile is unchanged
    cloudtruth!("config current")
        .assert()
        .success()
        .stdout(eq(orig_profile.name().as_str()));

    // create new environment map
    let env_map = hashmap! {
        CT_PROFILE => test_profile.name().as_str()
    };

    // check current profile with new environment
    cloudtruth!("config current")
        .env_clear()
        .envs(&env_map)
        .assert()
        .success()
        .stdout(ne(orig_profile.name().as_str()).and(contains(test_profile.name())));

    // Check all profile parameters

    let profile_json = cloudtruth!("config current -sf json")
        .env_clear()
        .envs(&env_map)
        .assert()
        .success()
        .parse_profile_parameters();

    let expected_param_names = vec![
        "Profile",
        "API key",
        "Organization",
        "User",
        "Role",
        "Project",
        "Environment",
    ];
    let actual_param_names = profile_json.param_names();
    assert_eq!(expected_param_names, actual_param_names);

    let param = profile_json.find_param("Profile");
    assert_eq!(param.value, test_profile.name().as_str());
    assert_eq!(param.source, "shell");

    let param = profile_json.find_param("API key");
    assert_eq!(param.value, "bogus-key-value");
    assert_eq!(param.source, format!("profile ({test_profile})"));

    let param = profile_json.find_param("Project");
    assert_eq!(param.value, test_proj.name().as_str());
    assert_eq!(param.source, format!("profile ({test_profile})"));

    let param = profile_json.find_param("Environment");
    assert_eq!(param.value, test_env.name().as_str());
    assert_eq!(param.source, format!("profile ({test_profile})"));

    let param = profile_json.find_param("Organization");
    assert_eq!(param.value, "");
    assert_eq!(param.value, "");

    let param = profile_json.find_param("User");
    assert_eq!(param.value, "");
    assert_eq!(param.value, "");

    let param = profile_json.find_param("Role");
    assert_eq!(param.value, "");
    assert_eq!(param.value, "");

    // Extended profile configuration
    let env_map = hashmap! {
        CT_PROFILE => test_profile.name().as_str(),
        CT_REST_SUCCESS => "a,b,c,d",
        CT_REST_DEBUG => "false",
        CT_REST_PAGE_SIZE => "9",
        CT_REQ_TIMEOUT => "33"
    };

    let profile_json = cloudtruth!("config current -xsf json")
        .env_clear()
        .envs(&env_map)
        .assert()
        .success()
        .parse_profile_parameters();

    let expected_param_names = vec![
        "Profile",
        "API key",
        "Organization",
        "User",
        "Role",
        "Project",
        "Environment",
        "CLI version",
        "Server URL",
        "Request timeout",
        "REST debug",
        "REST success",
        "REST page size",
        "Accept Invalid Certs",
    ];
    let actual_param_names = profile_json.param_names();
    assert_eq!(expected_param_names, actual_param_names);

    let param = profile_json.find_param("REST debug");
    assert_eq!(param.value, "false");
    assert_eq!(param.source, "shell");

    let param = profile_json.find_param("REST success");
    assert_eq!(param.value, "a, b, c, d");
    assert_eq!(param.source, "shell");

    let param = profile_json.find_param("REST page size");
    assert_eq!(param.value, "9");
    assert_eq!(param.source, "shell");

    let param = profile_json.find_param("Request timeout");
    assert_eq!(param.value, "33");
    assert_eq!(param.source, "shell");

    // test with command line arguments

    let profile_json = cloudtruth!(
        "--api-key 'bogus-key-value' --profile '{test_profile}' config current -sf json"
    )
    .env_clear()
    .envs(&env_map)
    .assert()
    .success()
    .parse_profile_parameters();

    let param = profile_json.find_param("Profile");
    assert_eq!(param.value, test_profile.name().as_str());
    assert_eq!(param.source, "argument");

    let param = profile_json.find_param("API key");
    assert_eq!(param.value, "bogus-key-value");
    assert_eq!(param.source, "argument");

    let param = profile_json.find_param("Organization");
    assert_eq!(param.value, "");
    assert_eq!(param.value, "");

    let param = profile_json.find_param("User");
    assert_eq!(param.value, "");
    assert_eq!(param.value, "");

    let param = profile_json.find_param("Role");
    assert_eq!(param.value, "");
    assert_eq!(param.value, "");

    let deleted_profile = Profile::from_name(test_profile.name().clone());
    drop(test_profile);

    // when profile is not found, command succeeds without the bits from the config
    cloudtruth!("config current -s")
        .env_clear()
        .env(CT_PROFILE, deleted_profile.name().as_str())
        .assert()
        .success()
        .stdout(
            contains(deleted_profile.name().as_str())
                .and(not(contains("bogus-key-value")))
                .and(not(contains(test_env.name().as_str())))
                .and(not(contains(test_proj.name().as_str())))
                .and(not(contains!("profile ({deleted_profile})"))),
        );

    // use default environment variables
    let profile_json = cloudtruth!("config current -f json")
        .assert()
        .success()
        .parse_profile_parameters();
    let param = profile_json.find_param("API key");
    assert_eq!("*****", param.value);
    let param = profile_json.find_param("Organization");
    assert_ne!("", param.value);
    assert_eq!("API key", param.source);
    let param = profile_json.find_param("User");
    assert_ne!("", param.value);
    assert_eq!("API key", param.source);
    let param = profile_json.find_param("Role");
    assert!(["owner", "admin", "contrib"].contains(&param.value.as_str()));
    assert_eq!("API key", param.source);
}
